### PR TITLE
Enum.fetch/2: Optimized for ranges

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

**Comment by eksperimental** [comment]
I would merge this after https://github.com/elixir-lang/elixir/pull/4950 is merged, and probably will have to be rebased

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -2513,6 +2539,35 @@ defmodule Enum do
     lower_limit + :rand.uniform(upper_limit - lower_limit + 1) - 1
   end
 
+  # Returns a range in ascending order, and with index >= 0
+  @spec normalize_range(Range.t, integer) ::
+    :error |
+    {:ok, {Range.t, index, count}} when index: integer, count: pos_integer
```
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: I found a spec for this private func. helpful

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -717,6 +717,32 @@ defmodule Enum do
   """
   @spec fetch(t, integer) :: {:ok, element} | :error
   def fetch(enumerable, index)
+  def fetch(first.._, 0),
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Still not a big fun of such small clauses specially given the common case should be fast enough.

**Comment by josevalim** [comment]
@eksperimental I haven't tried the code below but what do you think about it:

```elixir
defmodule A do
  def fetch_range(first..last, index) when first <= last and index >= 0 do
    item = first + index
    if item > last, do: :error, else: {:ok, item}
  end

  def fetch_range(first..last, index) when first <= last  do
    item = last + index + 1
    if item < first, do: :error, else: {:ok, item}
  end

  def fetch_range(first..last, index) when index >= 0 do
    item = first - index
    if item < last, do: :error, else: {:ok, item}
  end

  def fetch_range(first..last, index) do
    item = last - index - 1
    if item > first, do: :error, else: {:ok, item}
  end
end
```

I found listing all 4 scenarios without going throw normalization to be clearer. It also seems more concise than the current implementation. My recommendation, along side the the previous PR, is to break the range and list implementations into fetch_range and fetch_list and avoid adding all of those cases directly into fetch.

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: I would like to do some benchmarks as I haven't done none yet. Let's see what's the performance like ;-D

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: benchmarks showed it's better not to optimize it.

**Comment by eksperimental** [comment]
tests have been rewritten.
merge please if everything is fine, I will rebase the other PR about fetch.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**